jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EagerContentHandler.java

```diff
@@ -0,0 +1,618 @@
+         *                      for framing overheads in the transport. Since the content is retained rather than copied, any
+         *                      framing data is also retained in the IO buffer.
+         * @param reject if {@code true}, then if {@code maxRetainBytes} is exceeded, the request is rejected with a
+         *               {@link HttpStatus#PAYLOAD_TOO_LARGE_413} response.
+         */
+        public RetainedContentLoaderFactory(long maxRetainedBytes, int frameOverhead, boolean reject)
+        {
+            _maxRetainedBytes = maxRetainedBytes;
+            _frameOverhead = frameOverhead;
+            _reject = reject;
+        }
+
+        @Override
+        public String getApplicableMimeType()
+        {
+            return null;
+        }
+
+        @Override
+        public ContentLoader newContentLoader(String contentType, String mimeType, Handler handler, Request request, Response response, Callback callback)
+        {
+            return new RetainedContentLoader(handler, request, response, callback, _maxRetainedBytes, _frameOverhead, _reject);
+        }
+
+        /**
+         * Delay dispatch until all content or an effective buffer size is reached
+         */
+        public static class RetainedContentLoader extends ContentLoader implements Invocable.Task
+        {
+            private final Deque<Content.Chunk> _chunks = new ArrayDeque<>();
+            private final long _maxRetainedBytes;
+            private final int _chunkOverhead;
+            private final boolean _reject;
+            private long _estimatedSize;
+
+            /**
+             * @param handler The next handler
+             * @param request The delayed request
+             * @param response The delayed response
+             * @param callback The delayed callback
+             * @param maxRetainedBytes The maximum size to buffer before dispatching to the next handler;
+             *                or -1 for a heuristically determined default
+             * @param frameOverhead The bytes to account for per chunk when calculating the size; or -1 for a heuristic.
+             * @param reject If {@code true} then requests are rejected if the content is not complete before maxRetainedBytes.
+             */
+            public RetainedContentLoader(Handler handler, Request request, Response response, Callback callback, long maxRetainedBytes, int frameOverhead, boolean reject)
+            {
+                super(handler, request, response, callback);
+                _maxRetainedBytes = maxRetainedBytes < 0
+                    ? Math.max(1, request.getConnectionMetaData().getConnector().getConnectionFactory(HttpConnectionFactory.class).getInputBufferSize() - 1500)
+                    : maxRetainedBytes;
+                _chunkOverhead = frameOverhead < 0
+                    ? (request.getConnectionMetaData().getHttpVersion() == HttpVersion.HTTP_2 ? 9 : 8)
```

Comment:
Also for HTTP/3, the framing depends on the content size, but it can only be at most 9 bytes.

So perhaps:

```suggestion
                    ? (request.getConnectionMetaData().getHttpVersion().getVersion() <= HttpVersion.HTTP_1_1.getVersion() ? 8 : 9)
```